packages/flutter/lib/src/painting/shader_warm_up.dart

```diff
@@ -0,0 +1,144 @@
+/// compilations from animation time to startup time. By default, a
+/// [DefaultShaderWarmUp] is used. Create a custom [ShaderWarmUp] subclass to
+/// replace it if needed.
+///
+/// This warm up needs to be run on each individual device because the shader
+/// compilation depends on the specific GPU hardware and driver a device has. It
+/// can't be pre-computed during the Flutter engine compilation as the engine is
+/// device agnostic.
+abstract class ShaderWarmUp {
+  /// Allow const constructors for subclasses.
+  const ShaderWarmUp();
+
+  /// Width of the image to be drawn.
+  int get width;
+
+  /// Height of the image to be drawn.
+  int get height;
+
+  /// Trigger draw operations on a given canvas to warm up GPU shader
+  /// compilation cache.
+  void warmUpOnCanvas(Canvas canvas);
+
+  /// Construct an offscreen image of size [width]x[height], and execute
+  /// [warmUpOnCanvas] on a canvas associated with that image.
+  void execute() {
+    final PictureRecorder recorder = PictureRecorder();
+    final Canvas canvas = Canvas(recorder);
+
+    warmUpOnCanvas(canvas);
+
+    final Picture picture = recorder.endRecording();
+    final TimelineTask shaderWarmUpTask = TimelineTask();
+    shaderWarmUpTask.start('Warm-up shader');
+    picture.toImage(width, height).then((Image image) {
+      shaderWarmUpTask.finish();
+    });
+  }
+}
+
+/// Default way of warming up Skia shader compilations.
+///
+/// The draw operations being warmed up here are decided according to Flutter
+/// engineers' observation and experience based on the apps and the performance
+/// issues seen so far.
+///
+/// To decide which draw operations to be added to your custom warm up process,
+/// try capture an skp using `flutter screenshot --observatory- port=<port>
+/// --type=skia` and analyze it with https://debugger.skia.org. Alternatively,
+/// one may run the app with `flutter run --trace-skia` and then examine the GPU
+/// thread in the observatory timeline to see which Skia draw operations are
+/// commonly used, and which shader compilations are causing janks.
+///
+/// The size of [width]=[height]=1000 here are picked arbitrarily. The only goal
+/// is to have them not too far away from the actual resolution (very few mobile
+/// devices seem to have a resolution that is 10x larger or 10x smaller than
+/// 1000x1000). The size should have little effect on the shader warm-up. A
+/// custom shader warm up can pick another arbitrary number that's close to the
+/// targeted devices.
+class DefaultShaderWarmUp extends ShaderWarmUp {
+  /// Allow [DefaultShaderWarmUp] to be used as the default value of parameters.
+  const DefaultShaderWarmUp();
+
+  @override
+  int get width => 1000;
+
+  @override
+  int get height => 1000;
+
+  /// Trigger common draw operations on a canvas to warm up GPU shader
+  /// compilation cache.
+  @override
+  void warmUpOnCanvas(Canvas canvas) {
+    final RRect rrect = RRect.fromLTRBXY(20, 20, 60, 60, 10, 10);
```

Comment:
please use `.0` on double literals so it's clear that they're doubles. also applies in other places below.